### PR TITLE
Introduce pre-process function for resources

### PR DESCRIPTION
**Pull request by slaskawi**
This Pull Request introduces a pre-process function to the resources. This gives a caller an opportunity to alter them after decoding. 

Proof PR: https://github.com/openshift/cert-manager-operator/pull/44

### CODE REVIEW

**Comment by slaskawi** [comment]
@sttts @s-urbaniak @stlaz Can I ask for a review?

File: pkg/operator/resource/resourceapply/generic.go [hunk 1]
```diff
@@ -371,6 +392,16 @@ func DeleteAll(ctx context.Context, clients *ClientHolder, recorder events.Recor
 	return ret
 }
 
+func DeleteAll(ctx context.Context, clients *ClientHolder, recorder events.Recorder, manifests AssetFunc, files ...string) []ApplyResult {
+	return DeleteAllWithPreprocess(ctx, clients, recorder, manifests, defaultPreprocessFunc(), files...)
+}
+
+func defaultPreprocessFunc() func(ctx context.Context, object runtime.Object) (runtime.Object, error) {
```
Comment (s-urbaniak) [inline pkg/operator/resource/resourceapply/generic.go:399]: naming suggestion: `nopPreprocess`

File: pkg/operator/resource/resourceapply/generic.go [hunk 2]
```diff
@@ -45,6 +45,10 @@ func init() {
 	utilruntime.Must(policyv1.AddToScheme(genericScheme))
 }
 
+// PreprocessResourcesFuncType gives the caller an opportunity to preprocess decoded resources
+// before applying them.
+type PreprocessResourcesFuncType func(ctx context.Context, object runtime.Object) (runtime.Object, error)
```
Comment (s-urbaniak) [inline pkg/operator/resource/resourceapply/generic.go:50]: naming suggestion: `PreprocessFunc`

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
a few naming suggestions, and generally: let's add unit tests for the introduced preprocessing.

File: pkg/operator/resource/resourceapply/generic.go (hunk 2 above)
Comment (slaskawi) [reply pkg/operator/resource/resourceapply/generic.go:50]: I'd like to leave it as is. It's symmetric with `PreprocessResourcesFuncType` in `static_resource_controller.go`.

File: pkg/operator/resource/resourceapply/generic.go (hunk 1 above)
Comment (slaskawi) [reply pkg/operator/resource/resourceapply/generic.go:399]: I'd like to leave it as is. It's symmetric with `defaultPreprocessFunc` in `static_resource_controller.go`.

File: pkg/operator/resource/resourceapply/generic.go [hunk 3]
```diff
@@ -242,8 +258,7 @@ func ApplyDirectly(ctx context.Context, clients *ClientHolder, recorder events.R
 	return ret
 }
 
-func DeleteAll(ctx context.Context, clients *ClientHolder, recorder events.Recorder, manifests AssetFunc,
-	files ...string) []ApplyResult {
+func DeleteAllWithPreprocess(ctx context.Context, clients *ClientHolder, recorder events.Recorder, manifests AssetFunc, preprocessFunction func(ctx context.Context, object runtime.Object) (runtime.Object, error), files ...string) []ApplyResult {
```
Comment (s-urbaniak) [inline pkg/operator/resource/resourceapply/generic.go:261]: same here, shouldn't this be:
```
func DeleteAllWithPreprocess(ctx context.Context, clients *ClientHolder, recorder events.Recorder, manifests AssetFunc, preprocessFunction PreprocessResourcesFuncType, files ...string) []ApplyResult {

```

File: pkg/operator/resource/resourceapply/generic.go [hunk 4]
```diff
@@ -102,6 +106,12 @@ func (c *ClientHolder) WithMigrationClient(client migrationclient.Interface) *Cl
 
 // ApplyDirectly applies the given manifest files to API server.
 func ApplyDirectly(ctx context.Context, clients *ClientHolder, recorder events.Recorder, cache ResourceCache, manifests AssetFunc, files ...string) []ApplyResult {
+	return ApplyDirectlyWithPreprocess(ctx, clients, recorder, cache, manifests, defaultPreprocessFunc(), files...)
+}
+
+// ApplyDirectlyWithPreprocess applies the given manifest files to API server (similarly to ApplyDirectly)
+// but also respects the pre-process function.
+func ApplyDirectlyWithPreprocess(ctx context.Context, clients *ClientHolder, recorder events.Recorder, cache ResourceCache, manifests AssetFunc, preprocessFunction func(ctx context.Context, object runtime.Object) (runtime.Object, error), files ...string) []ApplyResult {
```
[the remark this answers is not in the document]
Comment (slaskawi) [reply pkg/operator/resource/resourceapply/generic.go:114]: It depends...

We cannot freely use `PreprocessResourcesFuncType` freely here. If we did, the `static_resource_controller.go` would need to reuse this function type to implement `defaultStaticResourcesPreconditionsFunc` and we'd introduce a direct dependency between those two files. 

So the question is whether we want to introduce this dependency or not. At the moment, there is no such dependency and I wanted to keep it that way. The `static_resource_controller.go` can define its own `StaticResourcesPreconditionsFuncType` and as long as signatures match - it's fine with the compiler. 

Let me know if you still want it or not.

**Comment by slaskawi** [comment]
Rebased.

**Pull request closed without merging**